File: tools/glob.py
import fnmatch
import os
import time
from pathlib import Path


EXCLUDE_DIRS = {
    ".venv", "venv", ".git", "__pycache__", "node_modules", ".hg", ".svn",
    ".idea", ".vscode", ".tox", "build", "dist", ".next", ".turbo",
    "bin", "obj", "target", ".gradle", ".pytest_cache", ".mypy_cache",
    ".vs", ".vscode-test", ".cache", ".parcel-cache", "coverage", ".nuget",
}
NOISE_EXTS = {
    ".exe", ".dll", ".pdb", ".obj", ".lib", ".so", ".dylib", ".class",
    ".pyc", ".pyd", ".rar", ".zip", ".7z", ".gz", ".tar", ".nupkg",
    ".ico", ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".webp",
    ".mp3", ".mp4", ".avi", ".mkv", ".wav", ".ttf", ".otf", ".woff", ".woff2",
}
MAX_FILES = 20000
SCAN_TIMEOUT = 10
SUGGEST_TIMEOUT = 3


def list_files(pattern: str = "*", path: str = ".") -> str:
    """List files matching a glob pattern. Returns files sorted by modification time.

    Args:
        path: Directory to search in (default: current workspace)
        pattern: Glob pattern to match (default: "*")
    """
    try:
        search_path = Path(path).resolve()
        if not search_path.exists():
            return f"Error: Path not found: {path}"
        if not search_path.is_dir():
            return f"Error: Not a directory: {path}"

        return _fallback_glob(pattern or "*", search_path)
    except Exception as e:
        return f"Error listing files: {e}"


def _fallback_glob(pattern: str, search_path: Path) -> str:
    results = []
    deadline = time.time() + SCAN_TIMEOUT
    scanned = 0
    hidden = 0
    # An explicit pattern means the model asked for these files on purpose.
    explicit = pattern not in ("*", "*.*", "")

    # A pattern containing a separator addresses a path ("src/*.py"), which
    # never matches when only the bare file name is tested.
    norm_pattern = pattern.replace("\\", "/")
    path_scoped = "/" in norm_pattern

    for root, dirs, files in os.walk(str(search_path)):
        dirs[:] = [d for d in dirs if d not in EXCLUDE_DIRS]

        if time.time() > deadline:
            return f"Error: Scan timed out (> {SCAN_TIMEOUT}s). Narrow your search."

        for name in files:
            scanned += 1
            if scanned > MAX_FILES:
                return f"Error: Too many files ({MAX_FILES}+). Narrow your search."
            full = Path(root) / name
            try:
                rel = full.relative_to(search_path)
            except ValueError:
                continue
            if path_scoped:
                rel_posix = rel.as_posix()
                matched = fnmatch.fnmatch(rel_posix, norm_pattern)
                # "src/*.py" should also reach nested files, matching how the
                # model expects a directory-scoped glob to behave.
                if not matched and "**" not in norm_pattern:
                    matched = fnmatch.fnmatch(
                        rel_posix, norm_pattern.replace("/", "/*", 1)
                    )
            else:
                matched = fnmatch.fnmatch(name, pattern)
            if matched:
                if not explicit and Path(name).suffix.lower() in NOISE_EXTS:
                    hidden += 1
                    continue
                try:
                    stat = full.stat()
                except OSError:
                    continue
                results.append((rel, stat.st_size, stat.st_mtime))

    results.sort(key=lambda x: -x[2])

    if not results:
        return f"No files matching '{pattern}' in {search_path}"

    limit = 200 if "*" in pattern or "?" in pattern else 1000
    lines = [f"--- {len(results)} files matching '{pattern}' (newest first) ---"]
    for rel, size, mtime in results[:limit]:
        lines.append(
            f"{rel} ({_fmt_size(size)}, "
            f"{time.strftime('%Y-%m-%d %H:%M', time.localtime(mtime))})"
        )
    if len(results) > limit:
        lines.append(f"... and {len(results) - limit} more files")
    if hidden:
        lines.append(f"({hidden} binary/build files hidden)")
    return "\n".join(lines)


def suggest_files(basename: str, search_path: Path, max_results: int = 5) -> list[str]:
    """Find workspace files whose name matches a target basename (for 'did you mean').

    Matches on exact name, then on 'contains'. Ignores build/binary dirs.
    Returns relative paths, best matches first.
    """
    target = (basename or "").lower().replace("\\", "/").rstrip("/").split("/")[-1]
    if not target or target in (".", ".."):
        return []
    try:
        root_path = Path(search_path).resolve()
    except OSError:
        return []
    if not root_path.is_dir():
        return []

    exact = []
    contains = []
    # A shorter budget than the full scan: this runs on an error path, where a
    # ten-second stall reads like the agent has hung.
    deadline = time.time() + SUGGEST_TIMEOUT
    try:
        for root, dirs, files in os.walk(str(root_path)):
            dirs[:] = [d for d in dirs if d not in EXCLUDE_DIRS and not d.startswith(".")]
            if time.time() > deadline:
                break
            for name in files:
                low = name.lower()
                if low == target:
                    exact.append(str((Path(root) / name).relative_to(root_path)))
                elif target in low and Path(name).suffix.lower() not in NOISE_EXTS:
                    contains.append(str((Path(root) / name).relative_to(root_path)))
            if len(exact) >= max_results:
                break
    except (OSError, ValueError):
        return []
    exact.sort()
    contains.sort()
    return (exact + contains)[:max_results]


def _fmt_size(size: int) -> str:
    if size < 1024:
        return f"{size}B"
    if size < 1024 * 1024:
        return f"{size / 1024:.1f}KB"
    return f"{size / 1024 / 1024:.1f}MB"

File: internal/tools/glob.go
package tools

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// EXCLUDE_DIRS are directory names skipped during file walks (build/binary noise).
var EXCLUDE_DIRS = map[string]bool{
	".venv": true, "venv": true, ".git": true, "__pycache__": true,
	"node_modules": true, ".hg": true, ".svn": true, ".idea": true,
	".vscode": true, ".tox": true, "build": true, "dist": true,
	".next": true, ".turbo": true, "bin": true, "obj": true,
	"target": true, ".gradle": true, ".pytest_cache": true,
	".mypy_cache": true, ".vs": true, ".vscode-test": true,
	".cache": true, ".parcel-cache": true, "coverage": true, ".nuget": true,
}

// NOISE_EXTS are file extensions treated as build/binary noise.
var NOISE_EXTS = map[string]bool{
	".exe": true, ".dll": true, ".pdb": true, ".obj": true, ".lib": true,
	".so": true, ".dylib": true, ".class": true, ".pyc": true, ".pyd": true,
	".rar": true, ".zip": true, ".7z": true, ".gz": true, ".tar": true,
	".nupkg": true, ".ico": true, ".png": true, ".jpg": true, ".jpeg": true,
	".gif": true, ".bmp": true, ".webp": true, ".mp3": true, ".mp4": true,
	".avi": true, ".mkv": true, ".wav": true, ".ttf": true, ".otf": true,
	".woff": true, ".woff2": true,
}

const (
	maxFiles     = 20000
	scanTimeout  = 10 * time.Second
	maxListLimit = 1000
)

// ListFiles lists files matching a glob pattern, newest first.
func ListFiles(workspace, pattern, path string) string {
	base := resolveUnder(workspace, path)
	if base == "" {
		return fmt.Sprintf("Error: Path not found: %s", path)
	}
	info, err := os.Stat(base)
	if err != nil || !info.IsDir() {
		return fmt.Sprintf("Error: Not a directory: %s", path)
	}

	results, err := walkGlob(base, pattern)
	if err != nil {
		return err.Error()
	}
	if len(results) == 0 {
		return fmt.Sprintf("No files matching '%s' in %s", pattern, base)
	}

	limit := maxListLimit
	if strings.ContainsAny(pattern, "*?") {
		limit = 200
	}
	var b strings.Builder
	fmt.Fprintf(&b, "--- %d files matching '%s' (newest first) ---\n", len(results), pattern)
	for i, r := range results {
		if i >= limit {
			fmt.Fprintf(&b, "... and %d more files\n", len(results)-limit)
			break
		}
		fmt.Fprintf(&b, "%s (%s, %s)\n", r.rel, fmtSize(r.size),
			r.mtime.Format("2006-01-02 15:04"))
	}
	return strings.TrimRight(b.String(), "\n")
}

type globResult struct {
	rel   string
	size  int64
	mtime time.Time
}

func walkGlob(base, pattern string) ([]globResult, error) {
	normPattern := strings.ReplaceAll(pattern, "\\", "/")
	pathScoped := strings.Contains(normPattern, "/")
	explicit := pattern != "*" && pattern != "*.*" && pattern != ""

	deadline := time.Now().Add(scanTimeout)
	scanned := 0
	var results []globResult

	err := filepath.Walk(base, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if time.Now().After(deadline) {
			return fmt.Errorf("Error: Scan timed out (> %ds). Narrow your search.", int(scanTimeout.Seconds()))
		}
		if info.IsDir() {
			if p != base && EXCLUDE_DIRS[info.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		scanned++
		if scanned > maxFiles {
			return fmt.Errorf("Error: Too many files (%d+). Narrow your search.", maxFiles)
		}

		name := info.Name()
		var matched bool
		if pathScoped {
			rel, _ := filepath.Rel(base, p)
			relPosix := filepath.ToSlash(rel)
			matched, _ = filepath.Match(normPattern, relPosix)
			if !matched && !strings.Contains(normPattern, "**") {
				alt := strings.Replace(normPattern, "/", "/*", 1)
				matched, _ = filepath.Match(alt, relPosix)
			}
		} else {
			matched, _ = filepath.Match(pattern, name)
		}
		if !matched {
			return nil
		}

		if !explicit && NOISE_EXTS[strings.ToLower(filepath.Ext(name))] {
			return nil
		}

		rel, _ := filepath.Rel(base, p)
		results = append(results, globResult{rel: filepath.ToSlash(rel), size: info.Size(), mtime: info.ModTime()})
		return nil
	})
	if err != nil {
		return nil, err
	}

	sort.Slice(results, func(i, j int) bool { return results[i].mtime.After(results[j].mtime) })
	return results, nil
}

// SuggestFiles finds workspace files whose name matches a target basename,
// for "did you mean" hints. Returns relative paths, best matches first.
func SuggestFiles(workspace, target string, maxResults int) []string {
	if maxResults <= 0 {
		maxResults = 5
	}
	target = strings.ToLower(filepath.Base(strings.ReplaceAll(target, "\\", "/")))
	if target == "" {
		return nil
	}

	base := resolveUnder(workspace, ".")
	if base == "" {
		return nil
	}

	var exact, contains []string
	deadline := time.Now().Add(scanTimeout)
	_ = filepath.Walk(base, func(p string, info os.FileInfo, err error) error {
		if err != nil || time.Now().After(deadline) {
			return nil
		}
		if info.IsDir() {
			if p != base && EXCLUDE_DIRS[info.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		low := strings.ToLower(info.Name())
		rel, _ := filepath.Rel(base, p)
		rel = filepath.ToSlash(rel)
		if low == target {
			exact = append(exact, rel)
		} else if strings.Contains(low, target) && !NOISE_EXTS[strings.ToLower(filepath.Ext(info.Name()))] {
			contains = append(contains, rel)
		}
		return nil
	})

	sort.Strings(exact)
	sort.Strings(contains)
	merged := append(exact, contains...)
	if len(merged) > maxResults {
		merged = merged[:maxResults]
	}
	return merged
}

// resolveUnder resolves an absolute or workspace-relative path strictly under
// the workspace. Returns "" when the input is unusable.
func resolveUnder(workspace, p string) string {
	if p == "" {
		p = "."
	}
	if !filepath.IsAbs(p) {
		p = filepath.Join(workspace, p)
	}
	return filepath.Clean(p)
}

func fmtSize(size int64) string {
	switch {
	case size < 1024:
		return fmt.Sprintf("%dB", size)
	case size < 1024*1024:
		return fmt.Sprintf("%.1fKB", float64(size)/1024)
	default:
		return fmt.Sprintf("%.1fMB", float64(size)/1024/1024)
	}
}
